optaplanner-core/src/main/java/org/optaplanner/core/impl/partitionedsearch/queue/PartitionQueue.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.partitionedsearch.queue;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.optaplanner.core.impl.heuristic.selector.common.iterator.UpcomingSelectionIterator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Thread-safe
+ */
+public class PartitionQueue<Solution_> implements Iterable<PartitionChangeMove> {
+
+    protected final transient Logger logger = LoggerFactory.getLogger(getClass());
+
+    private BlockingQueue<PartitionChangedEvent<Solution_>> queue;
+    private Map<Integer, PartitionChangedEvent<Solution_>> moveEventMap; // Key is partIndex
+
+    // Only used by producers
+    private final Map<Integer, AtomicLong> nextEventIndexMap;
+
+    // Only used by consumer
+    private int openPartCount;
+    private final Map<Integer, Long> processedEventIndexMap; // Key is partIndex
+
+    public PartitionQueue(int partCount) {
+        // TODO partCount * 100 is pulled from thin air
+        queue = new ArrayBlockingQueue<>(partCount * 100);
+        moveEventMap = new ConcurrentHashMap<>(partCount);
+        // HashMap because it's never modified after creation
+        nextEventIndexMap = new HashMap<>(partCount);
+        for (int i = 0; i < partCount; i++) {
+            nextEventIndexMap.put(i, new AtomicLong(0));
+        }
```

Comment:
This is merely a suggestion, but maybe you can enforce immutability of `nextEventIndexMap` using `Collections.unmodifiableMap()` if you don't want to rely on `ConcurrentHashMap`